extensions/sql-database-projects/src/models/deploy/deployService.ts

```diff
@@ -0,0 +1,403 @@
+			return await task();
+		}
+	}
+
+	private async retry<T>(
+		name: string,
+		attempt: () => Promise<T>,
+		verify: (result: T) => Promise<ValidationResult>,
+		formatResult: (result: T) => Promise<string>,
+		numberOfAttempts: number = 10,
+		waitInSeconds: number = 2): Promise<T | undefined> {
+		for (let count = 0; count < numberOfAttempts; count++) {
+			this.logToOutput(constants.retryWaitMessage(waitInSeconds, name));
+			await new Promise(c => setTimeout(c, waitInSeconds * 1000));
+			this.logToOutput(constants.retryRunMessage(count, numberOfAttempts, name));
+
+			try {
+				let result = await attempt();
+				const validationResult = await verify(result);
+				const formattedResult = await formatResult(result);
+				if (validationResult.validated) {
+					this.logToOutput(constants.retrySucceedMessage(name, formattedResult));
+					return result;
+				} else {
+					this.logToOutput(constants.retryFailedMessage(name, formattedResult, validationResult.errorMessage));
+				}
+
+			} catch (err) {
+				this.logToOutput(constants.retryMessage(name, err));
+			}
+		}
+
+		return undefined;
+	}
+
+	private logToOutput(message: string): void {
+		if (this._outputChannel) {
+			this._outputChannel.appendLine(message);
+		}
+	}
+
+	private async createCommands(mssqlFolderPath: string, commandsFolderPath: string, dockerFilePath: string, startFilePath: string, imageLabel: string): Promise<void> {
+		// Create mssql folders if doesn't exist
+		//
+		await utils.createFolderIfNotExist(mssqlFolderPath);
+		await utils.createFolderIfNotExist(commandsFolderPath);
+
+		// Start command
+		//
+		await this.createFile(startFilePath, 'echo starting the container!');
+		if (os.platform() !== 'win32') {
+			await this.executeCommand(`chmod +x ${startFilePath}`);
+		}
+
+		// Create the Dockerfile
+		//
+		await this.createFile(dockerFilePath,
+			`
+FROM ${constants.dockerBaseImage}
+ENV ACCEPT_EULA=Y
+ENV MSSQL_PID=Developer
+LABEL ${imageLabel}
+RUN mkdir -p /opt/sqlproject
+COPY ${constants.mssqlFolderName}/${constants.commandsFolderName}/ /opt/commands
+RUN ["/bin/bash", "/opt/commands/start.sh"]
+`);
+	}
+
+	private async createFile(filePath: string, content: string): Promise<void> {
+		this.logToOutput(`Creating file ${filePath}, content:${content}`);
+		await fse.writeFile(filePath, content);
+	}
+
+	public async cleanDockerObjects(commandToGetObjects: string, commandsToClean: string[]): Promise<void> {
+		const currentIds = await this.executeCommand(commandToGetObjects);
+		if (currentIds) {
+			const ids = currentIds.split(/\r?\n/);
+			for (let index = 0; index < ids.length; index++) {
+				const id = ids[index];
+				if (id) {
+					for (let commandId = 0; commandId < commandsToClean.length; commandId++) {
+						const command = commandsToClean[commandId];
+						await this.executeCommand(`${command} ${id}`);
+
+					}
+				}
+			}
+		}
+	}
+
+	private async executeCommand(cmd: string, timeout: number = 5 * 60 * 1000): Promise<string> {
```

Comment:
As Benjin mentioned - this should definitely go in a common location. It's not specific to the deploy service at all